extensions/spark/kyuubi-extension-spark-3-5/src/main/scala/org/apache/kyuubi/sql/KyuubiEnsureRequirements.scala

```diff
@@ -0,0 +1,136 @@
+
+    // Get the indexes of children which have specified distribution requirements and need to have
+    // same number of partitions.
+    val childrenIndexes = requiredChildDistributions.zipWithIndex.filter {
+      case (UnspecifiedDistribution, _) => false
+      case (_: BroadcastDistribution, _) => false
+      case _ => true
+    }.map(_._2)
+
+    val childrenNumPartitions =
+      childrenIndexes.map(children(_).outputPartitioning.numPartitions).toSet
+
+    if (childrenNumPartitions.size > 1) {
+      // Get the number of partitions which is explicitly required by the distributions.
+      val requiredNumPartitions = {
+        val numPartitionsSet = childrenIndexes.flatMap {
+          index => requiredChildDistributions(index).requiredNumPartitions
+        }.toSet
+        assert(
+          numPartitionsSet.size <= 1,
+          s"$operator have incompatible requirements of the number of partitions for its children")
+        numPartitionsSet.headOption
+      }
+
+      // If there are non-shuffle children that satisfy the required distribution, we have
+      // some tradeoffs when picking the expected number of shuffle partitions:
+      // 1. We should avoid shuffling these children.
+      // 2. We should have a reasonable parallelism.
+      val nonShuffleChildrenNumPartitions =
+        childrenIndexes.map(children).filterNot(_.isInstanceOf[ShuffleExchangeExec])
+          .map(_.outputPartitioning.numPartitions)
+      val expectedChildrenNumPartitions =
+        if (nonShuffleChildrenNumPartitions.nonEmpty) {
+          if (nonShuffleChildrenNumPartitions.length == childrenIndexes.length) {
+            // Here we pick the max number of partitions among these non-shuffle children.
+            nonShuffleChildrenNumPartitions.max
+          } else {
+            // Here we pick the max number of partitions among these non-shuffle children as the
+            // expected number of shuffle partitions. However, if it's smaller than
+            // `conf.numShufflePartitions`, we pick `conf.numShufflePartitions` as the
+            // expected number of shuffle partitions.
+            math.max(nonShuffleChildrenNumPartitions.max, conf.defaultNumShufflePartitions)
+          }
+        } else {
+          childrenNumPartitions.max
+        }
+
+      val targetNumPartitions = requiredNumPartitions.getOrElse(expectedChildrenNumPartitions)
+
+      children = children.zip(requiredChildDistributions).zipWithIndex.map {
+        case ((child, distribution), index) if childrenIndexes.contains(index) =>
+          if (child.outputPartitioning.numPartitions == targetNumPartitions) {
+            child
+          } else {
+            val defaultPartitioning = distribution.createPartitioning(targetNumPartitions)
+            child match {
+              // If child is an exchange, we replace it with a new one having defaultPartitioning.
+              case ShuffleExchangeExec(_, c, _, advisoryPartitionSize) => ShuffleExchangeExec(
```

Comment:
Made a little change here, adding the `advisoryPartitionSize` parameter.

Reply:
it seems there are some more code changes in `EnsureRequirements`, e.g., storage partitioned join. I'm not very sure `KyuubiEnsureRequirements` still work. It would be better to copy and update related code.

Reply:
Which spark version is the current code copied from? I want to know what changes need to be made.

Reply:
It is from 3.1. We should copy 3.5..

Reply:
Thanks, I compared and found that this class has removed the following relevant code. I will try to copy it from spark3.5 and make changes.

![image](https://github.com/apache/kyuubi/assets/17894939/e7cfb4ae-e4d4-40fd-b9ed-86e291eb8582)

Reply:
I have regenerated `KyuubiEnsureRequirements` based on `EnsureRequirements` in spark 3.5 and made the following changes:

+ remove shuffle pruning
+ remove `reorderJoinPredicates` method
+ remove `requiredDistribution` parameter